rfcs/2021-02-04-5802-event-metadata.md

```diff
@@ -0,0 +1,206 @@
+
+```rust
+struct Event {
+    data: EventData,
+    metadata: EventMetadata,
+}
+
+enum EventData {
+    Log(LogEvent),
+    Metric(Metric),
+}
+```
+
+All of the methods on `EventData` will remain in place, but this will
+require additional handling for both the creation of new events (since
+the metadata will need to be filled in) and the "into" methods (`fn
+into_log` and `fn into_metric`). Implementations for `Deref` and
+`DerefMut` will simplify porting existing code to the new structure.
+
+### Vector Protocol
+
+The event protocol is used for both the transmission of events between
+Vector instances and the storage of events in disk buffers. This
+metadata is only relevant within a single instance of Vector, so should
+not be transmitted across the wire. However, for the latter, the
+protobuf definition will need to be expanded to include the metadata by
+adding the appropriate structures.
+
+```protobuf
+message EventWithMetadata {
+  EventWrapper data = 1;
+  EventMetadata metadata = 2;
+}
+
+message EventMetadata {
+  google.protobuf.Timestamp first_timestamp = 1;
+  repeated SourceEvent source_events = 2;
+  EventDeliveryStatus status = 3;
+}
+
+message SourceEvent {
+  google.protobuf.Timestamp timestamp = 1;
+  string source_name = 2;
+  string event_id = 3;
+}
+
+enum EventDeliveryStatus {
+  …TBD…
+}
+```
+
+### Visibility
+
+This metadata will not be visible to users, either through remap or JSON
+transforms, unless a use case can demonstrate the need for it. Given the
+above structure of the metadata, there are no user-modifiable parts.
+
+## Doc-level Proposal
+
+N/A
+
+## Rationale
+
+This addition is required to support the two features mentioned in the
+motivation. These provide users with:
+
+1. metrics about how long events are taking to reach their destination,
+   which can be used in meta-analysis about the observation framework
+   itself, and
+2. assurance that events are fully retained between when they are
+   generated and when they are stored and processed, but reducing the
+   number of ways they may be lost in transit.
+
+Once in place, this metadata may also enable additional features not yet
+envisioned. For example, the VRL compiler collects type information on
+fields, but this can be expensive for arrays. Transforms such as
+`add_fields` could build a field type cache to improve the performance
+of this feature.
+
+The introduction of `SmallVec` in the metadata structure is driven by
+the assumption that most events will have a single source throughout
+their lifetime. `SmallVec` is a data structure that can inline a (fixed)
+number of elements before allocating memory. By using this feature, this
+common case can avoid an extra allocation and improve data locality.
+
+## Prior Art
+
+N/A
+
+## Drawbacks
+
+- The event data is a critical component of Vector. Any changes to this
+  structure impacts virtually the entire system.
+- All transforms that combine or duplicate events will have extra
```

Comment:
I am curious about how this will be handled for the `lua` and `wasm` transforms where they are black boxes and the link between incoming event and outgoing event is unclear.

Could we describe how we will allow users to maintain the metadata when these transforms are used? I'm imagining something like instructing them to copy the metadata from the source event to the events they are emitting. This also means we may need to change the way that events are passed to/from these runtimes.

Reply:
I think we will have to punt on user scripts in either of those transforms. There are just too many ways they can modify the events to have Vector do anything sensible with the metadata automatically. I think we are left with two options:
1. Events "drop" when they enter a user transform and we throw away the statistics, or
2. The scripts now have a requirement to handle the metadata themselves, as you describe.
I'm not happy about either option, but I don't see another way out.

Reply:
Agreed, I think I prefer option 2 as many simple transformations using those runtimes just modify the incoming event and emit it. Like:

```toml
[transforms.my_transform_id]
  # General
  type = "lua" # required
  inputs = ["my-source-or-transform-id"] # required
  version = "2" # required

  # Hooks
  hooks.process = '''
  function (event, emit)
    event.log.field = "value" -- set value of a field
    event.log.another_field = nil -- remove field
    event.log.first, event.log.second = nil, event.log.first -- rename field
    -- Very important! Emit the processed event.
    emit(event)
  end
  '''
```

So, as long as the `event` going into lua has the metadata associated with it, it would be output out with it.

If they are doing anything more complicated, like combining events or fanning out, then we'd need to document how they can preserve the metadata. We can maybe output a warning message if an event comes out of one of those transforms without any metadata to prompt them to take a look.

Reply:
I'll make some notes to that effect.

Reply:
Notably, we'll need to think about wasm. Right now it just JSON encodes/decodes to pass the event back and forth. We may need to encode the metadata as a special key or augment the call interface to pass the metadata as a separate JSON parameter.

Reply:
There is no requirement, though, for these two script transforms to emit exactly one event for each source event. They may emit none, one, or multiple. I agree most of the time this will just be one-to-one, but we should at least document the implications of it not being so.